### ISSUE TITLE
Add method to guess the scitypes of a table

### ISSUE THREAD

**Issue by ablaom** [alan-turing-institute/ScientificTypes.jl#2]
Would output a dictionary `changes` on column names, whose values are the suggested scitype for the column with that name. The user could could then edit the dictionary and then effect the type coercions with `coerce(changes, data)`. 

So, if the new method is called `suggest_scitypes` (better name?) the workflow would be something like:

```julia
juliia> X = (book=["red", "white", "blue", "blue"], 
               number=[0, 1, 1, 0, 1],
               is_broken=[0, 1, 1, 0])

julia> changes = suggest_scitypes(X)
Dict{Symbol,Any} with 3 entries:
  :book => Multiclass
  :number => OrderedFactor
  :is_broken => OrderedFactor

julia> changes[:number] = :Count 
julia> X_fixed = coerce(changes, X)
```

**Comment by tlienart** [comment]
maybe `auto_scitypes` or `auto_types`? 

As a side note, I think it would be nice if this was called by default when a user loads data from file using a standard command (e.g. something that wraps around CSV.jl).

```julia
X = readcsv("path/to/file.csv") # default = auto_types
X = readcsv("path/to/file.csv"; auto_types=false) # if the user wants to do everything themselves
X = readcsv("path/to/file.csv"; types=(...)) # pre-specification and then no auto-types needed
```

**Comment by ablaom** [comment]
Yes, but there should be a base method that is agnostic to table type, ie work not just for DataFrames.

**Comment by tlienart** [comment]
Here's a possible draft (I can't dev `Scitypes` due to the ongoing release)

```julia
function suggest_scitype(type, col, nrows)
   # TODO: deal with missing/union (?)
   unique_vals  = unique(col)
   nunique_vals = length(unique_vals)
   # heuristic for whens the number of unique values seems "pretty low"
   # as compared to the number of rows.
   # Heuristic 1:
   #     there's more than 15 rows and ≤ 3 unique values ==> multiclass
   # Heuristic 2:
   #     there's less than 10% unique values out of the number of rows or
   #     there's fewer than 100 unique values with more than 1_000 rows
   #     => if a number, return OrderedFactor
   #     => if somethign else, return Multiclass
   # Finally if there are many unique values, just use the standard heuristic
   if nunique_vals ≤ 3 && nrows > 15
      # extremely likely this is a binary case
      return Multiclass{nunique_vals}
   elseif nunique_vals < min(0.1*nrows, 100)
      type <: Real && return OrderedFactor{nunique_vals}
      return Multiclass{nunique_vals}
   else
      type <: AbstractFloat  && return Continuous
      type <: Integer        && return Count
      # XXX to complete
      return Unknown
   end
end

function auto_types(X)
   sch = schema(X)
   suggested_types = Dict{Symbol,Found}()
   for (name, type, col) in zip(sch.names, sch.types, eachcolumn(X))
      guesses[name] = suggest_scitype(type, col, sch.nrows)
   end
   return suggested_types
end
```

let me know what you think

**Comment by ablaom** [comment]
Looks good to me. Could go in src/conventions/mlj/auto_types.jl or similar.

@tlienart Have added a dev branch and given you write access.

**Comment by tlienart** [comment]
closed by #4
